### ISSUE TITLE
invariant_booleans doesn't understand list manipulation (or that getters aren't invariant)

### ISSUE THREAD

**Issue by Hixie** [dart-lang/linter#1858]
Consider:

```dart
void test(List<double> x) {
  if (x.length == 1)
    return;
  x.add(1.0);
  if (x.length == 1) // bogus invariant_booleans error
    return;
  x.removeLast();
  if (x.length == 1) // bogus invariant_booleans error
    return;
}
```

More generally it doesn't seem to understand that getters aren't invariant:

```dart
class Test {
  int _value = 0;
  int get value	{
    _value += 1;
    return _value;
  }
}

void test(Test x) {
  if (x.value == 0)
    return;
  if (x.value == 0) // bogus invariant_booleans error
    return;
}
```

There's [a number of other invariant_booleans errors](https://github.com/dart-lang/sdk/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+invariant_booleans) on file.

**Comment by srawlins** [comment]
The real list of invariant_booleans bugs is [this list of 19](https://github.com/dart-lang/linter/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+invariant_booleans).

The maturity state of this lint rule has thus been marked "experimental". http://dart-lang.github.io/linter/lints/invariant_booleans.html

**Comment by jamesderlin** [comment]
I feel like this is equivalent to https://github.com/dart-lang/linter/issues/1384.  In general, it seems that the analyzer does not consider that function calls might mutate variables.  Or maybe it does but assumes that local variables can't be modified?

Here's yet another example that came up when I was trying to write some string parsing code:

```dart
final list = [1, 2, 3];
final iterator = list.iterator..moveNext();
if (iterator.current == 1) {
  iterator.moveNext();
  if (iterator.current == 1) { // bogus invariant_booleans error
    print('foo');
  }
}
```

**Comment by srawlins** [comment]
We've deprecated `invariant_booleans` and will not be fixing correctness bugs going forward.
